File: saw-data-access/src/main/sncr/datalake/DLConfiguration.java
package sncr.datalake;

import com.typesafe.config.Config;
import org.apache.hadoop.fs.FileSystem;
import org.apache.log4j.Logger;
import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.SQLContext;
import sncr.saw.common.config.SAWServiceConfig;

import java.io.IOException;
import java.util.Set;


/**
 * Created by srya0001 on 2/10/2017.
 */
public class DLConfiguration {

    private static DLConfiguration instance = new DLConfiguration();
    public DLConfiguration getInstance(){ return instance; }

    protected static JavaSparkContext jctx;
    protected static SparkContext ctx;

    protected static SparkConf sparkConf;
    protected static SQLContext sqlctx;


    public org.apache.hadoop.conf.Configuration getConfig() {
        return config;
    }
    public SQLContext getSqlContext(){return sqlctx;}
    public SparkContext getSC(){return ctx; }
    public FileSystem getFS() {
        return fs;
    }

    protected static org.apache.hadoop.conf.Configuration config;
    protected Set<System> sysParams;
    protected static FileSystem fs;

    private static final Logger logger = Logger.getLogger(DLConfiguration.class.getName());

    static  {

        Config cfg = SAWServiceConfig.spark_conf();

        logger.debug("Configure MAPR: ");
        config = new org.apache.hadoop.conf.Configuration();
        config.set("fs.defaultFS", "maprfs:///");
        try {
            fs = FileSystem.get(config);
        } catch (IOException e) {
            logger.error("Could not get FS: ", e);
        }
        logger.debug("Configure spark context: ");


        sparkConf = new SparkConf().setAppName("SAW-DataAccess");
        sparkConf.set("spark.master", cfg.getString("master"));
        sparkConf.set("spark.executor.memory", cfg.getString("executor.memory"));
        sparkConf.set("spark.cores.max", cfg.getString("cores.max"));
        sparkConf.set("driver.memory", cfg.getString("driver.memory"));
        jctx = new JavaSparkContext(sparkConf);
        ctx = jctx.sc();
        sqlctx = new SQLContext(ctx);
    }





    protected void finalize() {
        try {
            fs.close();
        } catch (IOException e) {
            logger.error("Could not close file system: ", e);
        }
        ctx.stop();
    }

}

File: saw-data-access/src/main/scala/sncr/datalake/DLConfiguration.scala
package sncr.datalake

import java.io.IOException

import com.typesafe.config.Config
import files.HFileOperations
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.spark.{SparkConf, SparkContext}
import org.slf4j.{Logger, LoggerFactory}
import sncr.saw.common.config.SAWServiceConfig

/**
  *  The object holds Spark execution and MapR filesystem configuration variables and
  *  methods.
  *  the initSpark method should be called in first place.
  *  The reason to keep Spark configuration step separate from object initialization <init>
  *    is to make it more flexible to use.
  */
object DLConfiguration {

  private val logger: Logger = LoggerFactory.getLogger("sncr.datalake.DLConfiguration")
  private var initialized: Boolean = false

  def getConfig: Configuration = DLConfiguration.config
  def getSC: SparkContext = DLConfiguration.ctx
  def getFS: FileSystem = DLConfiguration.fs
  def getSparkConfig : SparkConf = sparkConf

  /**
    * Spark context/configuration holders.
    */
  protected val cfg: Config = SAWServiceConfig.spark_conf
  protected var ctx : SparkContext = null
  protected var sparkConf : SparkConf = null



  logger debug "Configure MAPR: "
  protected val config = new Configuration
  config.set("fs.defaultFS", "maprfs:///")
  protected var fs: FileSystem = null
  try {
    fs = FileSystem.get(config)
  }
  catch {
    case e: IOException => {
      logger.error("Could not get FS: ", e)
    }
  }

  /**
    * Row limit is configured/hardcoded value
    */

  val rowLimit = if (cfg.hasPath ("sql-executor.inline-data-store-limit-rows") ) cfg.getInt ("sql-executor.inline-data-store-limit-rows") else 100
  val jarLocation = cfg.getString ("sql-executor.jar-location")
  val commonLocation = cfg.getString ("sql-executor.output-location")
  val semanticLayerTempLocation = cfg.getString ("sql-executor.semantic-layer-tmp-location")
  val defaultOutputType = cfg.getString ("sql-executor.output-type")

  val jarFiles = HFileOperations.listJarFiles(jarLocation, ".jar")
  logger debug s"Attach to Spark job the following jar files: ${jarFiles.mkString("[", ", ", "]")}"


  /**
    *  The method should be called to create and configure Spark context.
    * PLease note, method attaches to Spark context jars files that will be copied to
    * worker nodes.
    * it is recommended to store such jars in separate directory to avoid overhead network traffic and
    * even jar incompatibility issues.
    *
    */

  def initSpark(): Unit = {
    if (initialized) return
    logger info s"Spark settings: \n Master: ${cfg.getString ("master")} \n Worker memory: ${cfg.getString ("executor.memory")} \n  Cores: ${cfg.getString ("cores.max")} \n Driver memory: ${cfg.getString ("driver.memory")} "
    sparkConf = new SparkConf()
    sparkConf.setAppName ("SAW-DataAccess")
    sparkConf.set ("spark.master", cfg.getString ("master") )
    sparkConf.set ("spark.executor.memory", cfg.getString ("executor.memory") )
    sparkConf.set ("spark.cores.max", cfg.getString ("cores.max") )
    sparkConf.set ("driver.memory", cfg.getString ("driver.memory") )
    sparkConf.set ("spark.sql.inMemoryColumnarStorage.compressed", "true")
    sparkConf.set ("spark.sql.inMemoryColumnarStorage.batchSize", String.valueOf (rowLimit) )
    ctx = SparkContext.getOrCreate(sparkConf)
    jarFiles.foreach(f => ctx.addJar( jarLocation + Path.SEPARATOR + f))
    initialized = true
  }
}

